torch_glow/src/CachingGraphRunner.cpp

```diff
@@ -20,61 +20,128 @@
-  f = mod.createFunction(functionName);
-  std::vector<glow::Placeholder *> inputPlaceholders;
-  std::vector<glow::Placeholder *> outputPlaceholders;
+  std::unique_ptr<Module> module = llvm::make_unique<Module>();
+  Function *f = module->createFunction(info->functionName);
 
   RETURN_IF_ERR(PyTorchModelLoader::loadJITGraph(
-      *f, *graph, inputs, inputPlaceholders, outputPlaceholders,
+      *f, *graph, inputs, info->inputPlaceholders, info->outputPlaceholders,
       getPyTorchLoaderSettings()));
 
   glow::CompilationContext cctx;
-  executionEngine_.compile(cctx);
 
-  glow::PlaceholderBindings bindings;
-  for (size_t i = 0; i < inputs.size(); ++i) {
-    glow::Placeholder *ph = inputPlaceholders[i];
+  RETURN_IF_ERR(hostManager_->addNetwork(std::move(module), cctx));
+
+  perGlowGraphInfoMap.insert({hash, std::move(info)});
+
+  return perGlowGraphInfoMap[hash].get();
+}
+
+llvm::Error CachingGraphRunner::runGraphImpl(const PerGlowGraphInfo &info,
+                                             torch::jit::Stack &stack) {
+  DCHECK_EQ(getGraphHash(info.node, stack), info.hash)
+      << "Tried to run a graph with incompatible input shapes.";
+
+  size_t numInputs = info.inputPlaceholders.size();
+
+  auto inputs = torch::jit::last(stack, numInputs);
+
+  std::unique_ptr<ExecutionContext> ctx = llvm::make_unique<ExecutionContext>();
+  auto *bindings = ctx->getPlaceholderBindings();
+
+  for (size_t i = 0; i < numInputs; ++i) {
+    glow::Placeholder *ph = info.inputPlaceholders[i];
     glow::TypeRef ty = ph->getType();
     glow::Tensor t(inputs[i].toTensor().data_ptr(), ty);
-    bindings.insert(ph, std::move(t));
+    bindings->insert(ph, std::move(t));
   }
 
   std::vector<at::IValue> outputs;
-  for (auto *ph : outputPlaceholders) {
+  for (auto *ph : info.outputPlaceholders) {
     std::vector<int64_t> sizes;
     for (auto size : ph->dims()) {
       sizes.push_back(static_cast<int64_t>(size));
     }
     auto ptT = at::empty(sizes);
     glow::Tensor t(ptT.data_ptr(), ph->getType());
     outputs.push_back(std::move(ptT));
-    bindings.insert(ph, std::move(t));
+    bindings->insert(ph, std::move(t));
   }
 
-  executionEngine_.run(bindings);
+  auto err =
+      hostManager_->runNetworkBlocking(info.functionName, std::move(ctx));
 
   torch::jit::drop(stack, numInputs);
   for (auto &output : outputs) {
     auto var = torch::autograd::make_variable(output.toTensor());
     stack.push_back(at::IValue(var));
   }
 
+  return err;
+}
+
+llvm::Error CachingGraphRunner::runGraph(const torch::jit::Node *node,
+                                         torch::jit::Stack &stack) {
+  PerGlowGraphInfo *info;
+  ASSIGN_VALUE_OR_RETURN_ERR(info, loadGraphImpl(node, stack));
+  RETURN_IF_ERR(runGraphImpl(*DCHECK_NOTNULL(info), stack));
   return llvm::Error::success();
 }
+
+CachingGraphRunner::CachingGraphRunner() {
+  constexpr size_t numGlowDevices = 1;
+  constexpr const char *glowBackendName = "Interpreter";
```

Comment:
Why do we hardcode this?

Reply:
As opposed to a flag or what?

Reply:
Yeah, I think we should be able to test on other backend (CPU, hardware). Flag can be more convenient.

Reply:
Place it into PyTorchLoaderSettings struct.